### PR TITLE
Changed the depth of the markov chain, store all words instead of frequency

### PR DESCRIPTION
**Pull request by HowardStark**
The fork changes some core aspects of the text learning and generating process. It now accounts for punctuation, can allow for completely random generation, has a depth of two, and more.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -107,15 +108,21 @@ func handleCommand(bot *tgbotapi.BotAPI, update tgbotapi.Update) {
 	if update.Message.Command() == "test" {
 		msg := tgbotapi.NewMessage(update.Message.Chat.ID, "yo")
 		msg.ReplyToMessageID = update.Message.MessageID
-		bot.Send(msg)
+		_, sendErr := bot.Send(msg)
+		if sendErr != nil {
```
Comment (billybobjoeaglt) [inline main.go]: why check error? there isnt really any need

File: markov.go [hunk 2]
```diff
@@ -2,67 +2,76 @@ package main
...
+	"fmt"
 	"math/rand"
 	"os"
+	"regexp"
 	"strings"
 	"time"
 )
 
-var END = "@END@"
+var (
```
Comment (billybobjoeaglt) [inline markov.go]: End and Start should be constants and Capitalized properly EX: "End" and "Start"

File: markov.go [hunk 3]
```diff
@@ -2,67 +2,76 @@ package main
...
+	END         = "@END@"
+	START       = [2]string{"@START 1@", "@START 2@"}
+	punctuation = map[string]bool{
```
Comment (billybobjoeaglt) [inline markov.go]: Maybe use regexp?

File: markov.go [hunk 4]
```diff
@@ -2,67 +2,76 @@ package main
...
+		",": true,
+		"!": true,
+		"?": true,
+		";": true,
+		":": true,
+		"&": true,
+	}
+)
 
 func generateMarkovResponse(inputText string) string {
```
Comment (billybobjoeaglt) [inline markov.go]: Dont seed rand every time this is called

File: markov.go [hunk 5]
```diff
@@ -76,7 +85,7 @@ func loadDataset(path string) (DataMapType, error) {
 	if err != nil {
 		return nil, err
 	}
-	defer dataFile.Close()
+	defer fmt.Println(dataFile.Close())
```
Comment (billybobjoeaglt) [inline markov.go]: Use log.Println, not fmt

File: markov.go [hunk 6]
```diff
@@ -109,6 +118,11 @@ func preprocessText(text string) string {
 	return strings.ToLower(text)
 }
 
+func processText(text string) []string {
```
Comment (billybobjoeaglt) [inline markov.go]: regexp should be out of function so it doesnt compile every time

File: markov.go (hunk 2 above)
Comment (nacgarg) [inline markov.go]: haha mb

File: markov.go (hunk 4 above)
Comment (nacgarg) [inline markov.go]: also mb

File: markov.go (hunk 5 above)
Comment (nacgarg) [inline markov.go]: also mb

**Comment by billybobjoeaglt** [comment]
Also I would like to mention that using go arrays is something you want to avoid. Check out the [official go markov implementation](https://golang.org/doc/codewalk/markov/) for details on how they did it

File: markov.go [hunk 7]
```diff
@@ -2,67 +2,76 @@ package main
...
-			current += DataDict[currWord][keys[i]]
+		options, ok := DataDict[previousItems]
+		if !ok {
+			return response
+		}
+		nextItem := options[rand.Intn(len(options))]
+		if nextItem == END {
+			return response
+		}
+		if _, isPunctuation := punctuation[nextItem]; isPunctuation {
```
Comment (billybobjoeaglt) [inline markov.go]: use regexp

File: main.go [hunk 8]
```diff
@@ -32,6 +32,8 @@ func main() {
 
 	flag.Parse()
 
+	SplitRegex = regexp.MustCompile(`([\w'-]+|[.,!?;&])`)
```
Comment (billybobjoeaglt) [inline main.go]: Put this outside of a function. Like `var SplitRegex = regexp.MustCompile()`

File: markov.go [hunk 9]
```diff
@@ -42,8 +43,8 @@ func generateMarkovResponse(inputText string) string {
 	if _, ok := DataDict[previousItems]; !ok {
 		return "Error! I don't understand that =("
 	}
+	rand.Seed(time.Now().Unix())
```
Comment (billybobjoeaglt) [inline markov.go]: Seed in `init()` or create a new rand.New() every func call

**Pull request merged**